toolchain/check/import_cpp.cpp

```diff
@@ -2047,6 +2047,62 @@ static auto ImportNameDeclIntoScope(Context& context, SemIR::LocId loc_id,
                                                            access_kind);
 }
 
+// Returns true if the scope is the top `Cpp` scope.
+static auto IsTopCppScope(Context& context, SemIR::NameScopeId scope_id)
+    -> bool {
+  const SemIR::NameScope& name_scope = context.name_scopes().Get(scope_id);
+  return name_scope.is_cpp_scope() && !context.name_scopes()
+                                           .Get(name_scope.parent_scope_id())
+                                           .is_cpp_scope();
```

Comment:
Shouldn't this only be called for `Cpp` name scopes? I think the non-`Cpp` parent scope would always be the package scope, so maybe you can avoid the second `Get`, such as:

```suggestion
  CARBON_CHECK(name_scope.is_cpp_scope());
  return name_scope.parent_scope_id() == SemIR::NameScopeId::Package;
```

Reply:
IIUC, no.
This is called when we don't find the name in the given scope.
The scope could be `Cpp.MyNamespace`, so here we verify we only try to look for builtins in the `Cpp` scope and not in any scope below it.

Reply:
Can you please try out the suggested change locally? Is `ImportNameFromCpp` ever called when `is_cpp_scope` is false? When I test it on my machine, it works fine; so if it doesn't work, please add a failing test to prevent regressions.

> The scope could be `Cpp.MyNamespace`, so here we verify we only try to look for builtins in the `Cpp` scope and not in any scope below it.

A file's namespaces behave similarly to a tree, and `NameScopeId::Package` is the tree's root; it's the file's package. When importing another package, they are added to that tree structure as children beneath the root.


So in an example such as:

```
import Cpp inline '''
namespace MyNamespace {}
''';

fn F() {
  let long : Cpp.MyNamespace.long = 1;
}
```

We have a tree structure:

- Root: `NameScopeId::Package`
  - Child: `Cpp`
    - Child: `MyNamespace`
  - Child: `F`

Imports from other packages always have `NameScopeId::Package` as the parent. Because `package Cpp` is never directly declared, only imported, it will always be an imported package namespace beneath `NameScopeId::Package`. Because other C++ namespaces go beneath `Cpp` (directly or indirectly), their parent will never be `NameScopeId::Package`. That's why I expect an if-and-only-if comparison to work.

The reason I suggest the `name_scope.is_cpp_scope()` isn't really necessary is because `ImportNameFromCpp` seems to only be called when `is_cpp_scope` is true, [here](https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/check/name_lookup.cpp#L189). While I suggested it in a `CARBON_CHECK` because you're currently querying it, it feels like optional re-validation to me.

Reply:
Right, **sorry**, done.
I was misreading your previous comment.